File: src/main/scala/com/ajjpj/adiagram/render/shapes/ATextShape.scala
package com.ajjpj.adiagram.render.shapes

import com.ajjpj.adiagram.geometry.ARect
import com.ajjpj.adiagram.render.base.{PartialImage, TextStyle}
import com.ajjpj.adiagram.geometry.transform.Translation
import javafx.scene.canvas.GraphicsContext
import javafx.geometry.VPos
import javafx.scene.text.TextAlignment
import com.ajjpj.adiagram.ui.{AScreenPos, AScreenRect, Zoom}

/**
 * @author arno
 */
class ATextShape(val bounds: ARect, text: String, textStyle: TextStyle) extends AShape {
  //TODO generalize - different parts of the propText may be formatted differently, propText may be multi-line, ...

  def render(zoom: Zoom) = PartialImage.fromGc(bounds.topLeft, bounds, zoom, None, (gc, t) => {
    render(gc, t, zoom)
  })

  def render(gc: GraphicsContext, t: Translation, zoom: Zoom) {
    textStyle.applyTo(gc, zoom)
    gc.setTextBaseline(VPos.BOTTOM)

    gc.setTextAlign(textStyle.textAlignment)
    val tb = AScreenRect (t(bounds), zoom)
    val renderPos = AScreenPos(renderX(tb), renderY(tb, zoom))
    gc.fillText(text, renderPos.x, renderPos.y)
  }

  private def renderY(tb: AScreenRect, zoom: Zoom) = textStyle.vpos match {
    case VPos.TOP    => tb.topLeft.y + textStyle.height(zoom)
    case VPos.CENTER => tb.topLeft.y + tb.height/2 + textStyle.height(zoom)/2
    case VPos.BOTTOM => tb.topLeft.y + tb.height
    case _ => throw new IllegalArgumentException("only TOP, CENTER and BOTTOM are supported")
  }

  def renderX(tb: AScreenRect) = textStyle.textAlignment match {
    case TextAlignment.LEFT   => tb.topLeft.x
    case TextAlignment.CENTER => (tb.topLeft halfWayTo tb.topRight).x
    case TextAlignment.RIGHT  => tb.topRight.x
    case _ => throw new IllegalArgumentException("JUSTIFY not supported yet") //TODO justify
  }
}

File: src/main/kotlin/com/ajjpj/adiagram/render/shapes/ATextShape.kt
package com.ajjpj.adiagram.render.shapes

import com.ajjpj.adiagram.Zoom
import com.ajjpj.adiagram.geometry.APoint
import com.ajjpj.adiagram.geometry.ARect
import com.ajjpj.adiagram.geometry.AffineTransformation
import com.ajjpj.adiagram.render.AScreenPos
import com.ajjpj.adiagram.render.AScreenRect
import com.ajjpj.adiagram.render.PartialImage
import com.ajjpj.adiagram.render.style.TextStyle
import javafx.geometry.VPos
import javafx.scene.canvas.GraphicsContext
import javafx.scene.text.TextAlignment

data class ATextShape (override val bounds: ARect, val text: String, val textStyle: TextStyle): AShape {
    //TODO generalize - different parts of the propText may be formatted differently, propText may be multi-line, ...

    override val pos: APoint
        get() = bounds.topLeft

    override val renderBounds
            get() = bounds

    override fun render(zoom: Zoom) = PartialImage.fromGc(bounds.topLeft, bounds, zoom, null, fun (gc: GraphicsContext, t: AffineTransformation) {
        render(gc, t, zoom)
    })

    fun render(gc: GraphicsContext, t: AffineTransformation, zoom: Zoom) {
        textStyle.applyTo(gc, zoom)
        gc.setTextBaseline(VPos.BOTTOM)

        gc.setTextAlign(textStyle.textAlignment)
        val tb = AScreenRect (t(bounds), zoom)
        val renderPos = AScreenPos(renderX(tb), renderY(tb, zoom))
        gc.fillText(text, renderPos.x, renderPos.y)
    }

    private fun renderY(tb: AScreenRect, zoom: Zoom): Double = when (textStyle.vpos) {
        VPos.TOP    -> tb.topLeft.y + textStyle.height(zoom)
        VPos.CENTER -> tb.topLeft.y + tb.height/2 + textStyle.height(zoom)/2
        VPos.BOTTOM -> tb.topLeft.y + tb.height
        else        -> throw IllegalArgumentException("only TOP, CENTER and BOTTOM are supported")
    }

    private fun renderX(tb: AScreenRect): Double = when (textStyle.textAlignment) {
        TextAlignment.LEFT   -> tb.topLeft.x
        TextAlignment.CENTER -> (tb.topLeft.halfWayTo (tb.topRight)).x
        TextAlignment.RIGHT  -> tb.topRight.x
        else -> throw IllegalArgumentException("JUSTIFY not supported yet") //TODO justify
    }
}


